File: upf/python/clients/java/JClient/grpc_server_test/jclient_upf.py
from concurrent import futures
import grpc
import logging
import threading
import queue
import upf

from JClientUPF_pb2 import PlanMessage, PlanOrHRequest, State
import JClientUPF_pb2_grpc


class JClientUPFServicer(JClientUPF_pb2_grpc.JClientUPFServicer):

    def __init__(self):
        JClientUPF_pb2_grpc.JClientUPFServicer.__init__(self)
        self.lock = threading.Lock()
        self.lock.acquire() # Start acquired!
        self.state_evaluation = None
        self.queue = queue.Queue()

    def solve(self, request, context):
        actions = []
        for a in request.actions:
            actions.append(upf.Action(a.name, set(a.precondition), set(a.positiveEffect), set(a.negativeEffect)))
        problem = upf.Problem(actions, set(request.init), set(request.goal))
        plan = []
        with upf.Planner(request.planner) as p:
            plan = p.solve(problem)
        return PlanMessage(actions=plan)

    def solveWithHeuristic(self, request_iterator, context):
        threading.Thread(target=self._response_thread, args=(request_iterator,)).start()
        while True:
            x = self.queue.get()
            if x is None:
                break
            #print("Send: %s" % x)
            yield x

    def heuristic(self, state):
        self.queue.put(PlanOrHRequest(stateToEvaluate=State(state=state)))
        self.lock.acquire()
        return self.state_evaluation

    def _response_thread(self, request_iterator):
        for request in request_iterator:
            #print("Recv: %s" % request)
            if request.HasField("problem"):
                pm = request.problem
                actions = []
                for a in pm.actions:
                    actions.append(upf.Action(a.name, set(a.precondition), set(a.positiveEffect), set(a.negativeEffect)))
                problem = upf.Problem(actions, set(pm.init), set(pm.goal))
                threading.Thread(target=self._planner_thread, args=(pm.planner, problem)).start()
            else:
                assert request.HasField("stateEvaluation")
                self.state_evaluation = request.stateEvaluation
                self.lock.release()

    def _planner_thread(self, planner, problem):
        with upf.Planner(planner) as p:
            plan = p.solve(problem, self.heuristic)
            self.queue.put(PlanOrHRequest(plan=PlanMessage(actions=plan)))
            self.queue.put(None)


def serve():
  server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
  JClientUPF_pb2_grpc.add_JClientUPFServicer_to_server(
      JClientUPFServicer(), server)
  server.add_insecure_port('[::]:50052')
  server.start()
  server.wait_for_termination()

if __name__ == '__main__':
    logging.basicConfig()
    serve()

File: upf/cpp/clients/java/JClient/grpc_server_test/jclient_upf.cxx
#include <grpc/grpc.h>
#include <grpcpp/server.h>
#include <grpcpp/server_builder.h>
#include <grpcpp/server_context.h>
#include <grpcpp/security/server_credentials.h>

#include "JClientUPF.grpc.pb.h"

#include "upf.hxx"

using grpc::Server;
using grpc::ServerBuilder;
using grpc::ServerContext;
using grpc::ServerReader;
using grpc::ServerReaderWriter;
using grpc::ServerWriter;
using grpc::Status;
using eu::aiplan4eu::jclient::grpc::ActionMessage;
using eu::aiplan4eu::jclient::grpc::ProblemMessage;
using eu::aiplan4eu::jclient::grpc::ProblemOrHAnswer;
using eu::aiplan4eu::jclient::grpc::PlanMessage;
using eu::aiplan4eu::jclient::grpc::PlanOrHRequest;
using eu::aiplan4eu::jclient::grpc::JClientUPF;

upf::Action ConvertActionMessage(const ActionMessage& action_message)
{
  std::set<std::string> precondition;
  std::set<std::string> positive_effect;
  std::set<std::string> negative_effect;
  for (int i=0; i<action_message.precondition_size(); i++) {
    precondition.insert(action_message.precondition(i));
  }
  for (int i=0; i<action_message.positiveeffect_size(); i++) {
    positive_effect.insert(action_message.positiveeffect(i));
  }
  for (int i=0; i<action_message.negativeeffect_size(); i++) {
    negative_effect.insert(action_message.negativeeffect(i));
  }
  upf::Action action(action_message.name(), precondition, positive_effect, negative_effect);
  return action;
}

upf::Problem ConvertProblemMessage(const ProblemMessage& problem_message)
{
  std::vector<upf::Action> actions;
  std::set<std::string> init;
  std::set<std::string> goal;
  for (int i=0; i<problem_message.init_size(); i++) {
    init.insert(problem_message.init(i));
  }
  for (int i=0; i<problem_message.goal_size(); i++) {
    goal.insert(problem_message.goal(i));
  }
  for (int i=0; i<problem_message.actions_size(); i++) {
    actions.push_back(ConvertActionMessage(problem_message.actions(i)));
  }
  upf::Problem problem(actions, init, goal);
  return problem;
}

class JClientUPFImpl final : public JClientUPF::Service {
 public:

  Status solve(ServerContext* context, const ProblemMessage* request, PlanMessage* response)
  {
    if (auto plan = upf::solve(request->planner(), ConvertProblemMessage(*request))) {
      for (auto& s : *plan) {
        response->add_actions(s);
      }
    }
    return Status::OK;
  }

  Status solveWithHeuristic(ServerContext* context,
                            ServerReaderWriter<PlanOrHRequest, ProblemOrHAnswer>* stream)
  {

    return Status::OK;
  }

};


int main() {
  std::string server_address("localhost:50052");
  JClientUPFImpl service;

  ServerBuilder builder;
  builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
  builder.RegisterService(&service);
  std::unique_ptr<Server> server(builder.BuildAndStart());
  std::cout << "Server listening on " << server_address << std::endl;
  server->Wait();

  return 0;
}
